notebooks/tests/__init__.py

```diff
@@ -0,0 +1,125 @@
+#
+"""Unit tests for demo notebooks."""
+import os
+import subprocess
+import tarfile
+import shutil
+import site
+from typing import Optional
+
+import requests
+
+
+def _find_notebook(notebook_file: str) -> str:
+    path_to_notebook = os.path.join("notebooks", notebook_file)
+    if not os.path.exists(path_to_notebook):
+        path_to_notebook = os.path.join("..", path_to_notebook)
+    if not os.path.exists(path_to_notebook):
+        path_to_notebook = os.path.join("..", path_to_notebook)
+    if not os.path.exists(path_to_notebook):
+        raise Exception(
+            f"Could not find {notebook_file} in {os.path.abspath('.')} and {os.path.abspath('..')}"
+        )
+    return os.path.abspath(path_to_notebook)
+
+
+def _get_unzip(filename: str, url: str) -> None:
+    if not os.path.exists(filename):
+        response = requests.get(url, stream=True)
+        file = tarfile.open(fileobj=response.raw, mode="r|gz")
+        file.extractall(path=".")
+
+
+def _get_spark() -> None:
+    filename = 'spark-3.0.3-bin-hadoop2.7'
+    url = "https://downloads.apache.org/spark/spark-3.0.3/spark-3.0.3-bin-hadoop2.7.tgz"
+    _get_unzip(filename, url)
+    os.environ['SPARK_HOME'] = os.path.join(os.getcwd(), filename)
+
+
+def _get_hadoop() -> None:
+    filename = 'hadoop-2.10.1'
+    url = "https://downloads.apache.org/hadoop/common/hadoop-2.10.1/hadoop-2.10.1.tar.gz"
+    _get_unzip(filename, url)
+    os.environ['HADOOP_HOME'] = os.path.join(os.getcwd(), filename)
+
+
+def _copy_all_hadoop_jars_to_pyflink() -> None:
+    _get_hadoop()
+    if not os.getenv("HADOOP_HOME"):
+        raise Exception("The HADOOP_HOME env var must be set and point to a valid Hadoop installation")
+
+    pyflink_lib_dir = _find_pyflink_lib_dir()
+    for i, jar in enumerate(_jar_files()):
+        shutil.copy(jar, pyflink_lib_dir)
+    print(f"Copyied {i} HADOOP jar files into the pyflink lib dir at location {pyflink_lib_dir}")
+
+
+def _find_pyflink_lib_dir() -> Optional[str]:
+    for dir in site.getsitepackages():
+        package_dir = os.path.join(dir, "pyflink", "lib")
+        if os.path.exists(package_dir):
+            return package_dir
+    return None
+
+
+def _jar_files() -> str:
+    for root, _, files in os.walk(os.getenv("HADOOP_HOME")):
+        for file in files:
+            if file.endswith(".jar"):
+                yield os.path.join(root, file)
+
+
+def _get(filename, url) -> None:
+    if os.path.exists(filename):
+        return filename
+    r = requests.get(url)
+    with open(filename, 'wb') as f:
+        f.write(r.content)
+
+
+def fetch_nessie() -> str:
+    runner = "nessie-quarkus-runner"
+    import pynessie
+    version = pynessie.__version__
+    url = "https://github.com/projectnessie/nessie/releases/download/nessie-{}/nessie-quarkus-{}-runner".format(version, version)
+    _get(runner, url)
+    os.chmod(runner, 0o777)
+    return runner
+
+
+def fetch_iceberg_flink() -> str:
+    filename = "iceberg-flink-runtime-0.12.0.jar"
```

Comment:
My thinking is these versions are all hard coded because they need specific effort to upgrade them, including ensuring correct url below and such. We could abstract that all away but its noisy for not much reason. There isn't much duplication of version numbers or names though

Reply:
Also note, these are all dependent on the iceberg 0.12.0 release. Which should happen imminently. While we are waiting for that release I am pulling jars from the RC candidates repo

Reply:
so I guess when we want to upgrade to a newer version, we would need to modify stuff in this file here right?

Reply:
yes, I have extracted the versions out to make it easier